File: zig/src/nfq_dns_filter.zig
const std = @import("std");
const netfilter = @cImport({
    @cInclude("libnetfilter_queue/libnetfilter_queue.h");
    @cInclude("linux/netfilter.h");
});
const netinet = @cImport({
    @cInclude("netinet/ip.h");
    @cInclude("netinet/tcp.h");
    @cInclude("netinet/udp.h");
});
const dns = @import("dns.zig");
const logger = @import("logger.zig");

const QueueHandle = ?*netfilter.nfq_handle;
const Queue = ?*netfilter.nfq_q_handle;

var blocklist: std.ArrayList([]const u8) = undefined;

fn callback(queue: Queue, nfmsg: ?*netfilter.nfgenmsg, nfa: ?*netfilter.nfq_data, data: ?*anyopaque) callconv(.C) c_int {
    _ = nfmsg;
    _ = data;

    var id: u32 = 0;
    const packet_hdr = netfilter.nfq_get_msg_packet_hdr(nfa);

    if (packet_hdr == null) {
        logger.log("Error: Failed to get packet header\n", .{});
        return netfilter.NF_DROP;
    }
    id = std.mem.bigToNative(u32, packet_hdr.*.packet_id);

    // Get the packet payload
    var payload: [*c]u8 = undefined;
    var payload_len: c_int = 0;
    payload_len = netfilter.nfq_get_payload(nfa, &payload);

    if (payload_len < 0) {
        logger.log("Error: Failed to get packet payload\n", .{});
        return netfilter.NF_DROP;
    }

    // Check the IP protocol
    const ip_header: *const dns.ip = @ptrCast(@alignCast(payload));
    const ip_header_len = (ip_header.ip_vhl & 0x0F) << 2; //ip_header.ip_hl << 2;
    const transport_payload: []u8 = payload[ip_header_len..@intCast(payload_len)];

    // UDP DNS
    if (ip_header.ip_p == netinet.IPPROTO_UDP) {
        const dns_payload = transport_payload[@sizeOf(dns.udphdr)..];

        // Decode the QNAME
        var buffer: [256]u8 = undefined;
        const qname = dns.decodeQname(dns_payload[0..], &buffer) catch |err| {
            logger.log("Error decoding QNAME: {}\n", .{err});
            return netfilter.nfq_set_verdict(queue, id, netfilter.NF_ACCEPT, 0, null);
        };

        logger.log("UDP DNS Packet handled (ID: {}), Domain: {s}\n", .{ id, qname });

        // check if domain is in blocklist
        for (blocklist.items) |domain| {
            if (std.mem.eql(u8, qname, domain)) {
                logger.log("Domain {s} is blocked. Dropping packet.\n", .{qname});
                return netfilter.NF_DROP;
            }
        }
    }
    // TCP DNS
    else if (ip_header.ip_p == netinet.IPPROTO_TCP) {
        // Most DNS traffic is UDP so we won't implement this for now.
        logger.log("TCP Packet handled (ID: {})\n", .{id});
    }

    // Accept all packets
    return netfilter.nfq_set_verdict(queue, id, netfilter.NF_ACCEPT, 0, null);
}

fn loadBlocklist(allocator: std.mem.Allocator, file_path: []const u8) !std.ArrayList([]const u8) {
    var list = std.ArrayList([]const u8).init(allocator);
    const file = try std.fs.cwd().openFile(file_path, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();

    var line_buffer: [256]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&line_buffer, '\n')) |line| {
        // skip any comments or empty lines
        if (line.len == 0 or line[0] == '#') continue;

        const domain = std.mem.trim(u8, line, " \r\n"); // trim whitespace

        try list.append(try allocator.dupe(u8, domain));
    }

    return list;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // command line args
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len < 2) {
        logger.log("Usage: {s} <blocklist_file>\n", .{args[0]});
        std.process.exit(1);
    }

    // load blocklist
    blocklist = try loadBlocklist(allocator, args[1]);
    defer {
        for (blocklist.items) |domain| {
            allocator.free(domain);
        }
        blocklist.deinit();
    }

    var h: ?*netfilter.nfq_handle = undefined;
    var qh: ?*netfilter.nfq_q_handle = undefined;
    var fd: c_int = undefined;
    var buf: [4096]u8 = undefined;
    var rv: c_int = undefined;

    h = netfilter.nfq_open();
    if (h == null) {
        logger.log("Error during nfq_open()\n", .{});
        std.process.exit(1);
    }
    defer _ = netfilter.nfq_close(h);

    // Unbind existing nf_queue handler for AF_INET (if any)
    if (netfilter.nfq_unbind_pf(h, netfilter.AF_INET) < 0) {
        logger.log("Error during nfq_unbind_pf()\n", .{});
        std.process.exit(1);
    }

    // Bind nfnetlink_queue as the handler for AF_INET
    if (netfilter.nfq_bind_pf(h, netfilter.AF_INET) < 0) {
        logger.log("Error during nfq_bind_pf()\n", .{});
        std.process.exit(1);
    }

    // Create a new queue handle
    qh = netfilter.nfq_create_queue(h, 0, callback, null);
    if (qh == null) {
        logger.log("Error during nfq_create_queue()\n", .{});
        std.process.exit(1);
    }
    defer _ = netfilter.nfq_destroy_queue(qh);

    // Set the mode to copy packet data to userspace
    if (netfilter.nfq_set_mode(qh, netfilter.NFQNL_COPY_PACKET, 0xffff) < 0) {
        logger.log("Error during nfq_set_mode()\n", .{});
        std.process.exit(1);
    }

    // Get the file descriptor associated with the queue
    fd = netfilter.nfq_fd(h);

    // Main loop to process packets
    while (true) {
        rv = @as(c_int, @intCast(std.c.recv(fd, &buf, buf.len, 0)));
        if (rv >= 0) {
            // Handle the packet
            _ = netfilter.nfq_handle_packet(h, &buf, rv);
        } else {
            logger.log("Error during recv()\n", .{});
            break;
        }
    }
}

File: go/nfq_dns_filter.go
//go:build dnsfilter

package main

import (
	"context"
	"math/rand"
	"os"
	"os/signal"
	"syscall"
	"time"

	nfqueue "github.com/florianl/go-nfqueue"
	"github.com/mdlayher/netlink"

	"nfq_go_router/logger"
)

var nf *nfqueue.Nfqueue
var blocklist = []string{
    "example.com",
}
var ignoreDNSProbability = 0.95
var rng = rand.New(rand.NewSource(time.Now().UnixNano()))


func isBlocked(qname string) bool {
	for _, blocked := range blocklist {
		if qname == blocked || (len(qname) > len(blocked) && qname[len(qname)-len(blocked)-1:] == "."+blocked) {
			return true
		}
	}
	return false
}

func callback(a nfqueue.Attribute) int {
	// dereference packet info
	id := *a.PacketID
	payload := *a.Payload

	// payload[0] is a byte contains IP version and header length
	// the lower 4 bits (payload[0] & 0x0F) represent the number of 32-bit words in IP header
	// multiply by 4 (<< 2) to get length in bytes (32-bit words = 4 byte words)
	ipHeaderLen := int((payload[0] & 0x0F) << 2)

	// check if its a UDP DNS packet
	isMinUdpSize := len(payload) >= ipHeaderLen+8
	isUdp := payload[9] == 17
	if isMinUdpSize && isUdp {
		udpHeader := payload[ipHeaderLen : ipHeaderLen+8]
		dstPort := int(udpHeader[2])<<8 | int(udpHeader[3])

		// check if destination port is 53 (DNS)
		if dstPort == 53 {
			if rng.Float64() < ignoreDNSProbability {
				// Accept 95% of DNS packets without checking
				logger.Log("UDP DNS packet randomly accepted (ID: %d)\n", id)
				nf.SetVerdict(id, nfqueue.NfAccept)
				return 0
			}

			dnsPayload := payload[ipHeaderLen+8:]
			// Decode the QNAME
			buffer := make([]byte, 256)
			qnameBytes, err := decodeQname(dnsPayload, buffer)
			if err != nil {
				logger.Log("Error decoding QNAME (ID: %d): %v\n", id, err)
				nf.SetVerdict(id, nfqueue.NfAccept)
				return 0
			}

			qname := string(qnameBytes) // Convert []byte to string
			if isBlocked(qname) {
				logger.Log("UDP DNS packet blocked (ID: %d), Domain: %s\n", id, qname)
				nf.SetVerdict(id, nfqueue.NfDrop)
				return 0
			}

			logger.Log("UDP DNS packet accepted (ID: %d), Domain: %s\n", id, qname)
		}
	} else {
		logger.Log("Data packet handled (ID: %d)\n", id)
	}

	// Accept all other packets
	nf.SetVerdict(id, nfqueue.NfAccept)
	return 0
}

func main() {
	config := nfqueue.Config{
		NfQueue:      0,                       // Queue number
		MaxPacketLen: 0xFFFF,                  // Maximum packet length
		MaxQueueLen:  0xFFFF,                  // Maximum queue length
		Copymode:     nfqueue.NfQnlCopyPacket, // Copy packet data to userspace
		WriteTimeout: 15 * time.Millisecond,
	}

	// Open nfq
	var err error
	nf, err = nfqueue.Open(&config)
	if err != nil {
		logger.Log("could not open nfqueue socket:", err)
		return
	}
	defer nf.Close()

	// Enable NoENOBUFS option -- get the error `netlink receive: recvmsg: no buffer space available` if not enabled
	if err := nf.SetOption(netlink.NoENOBUFS, true); err != nil {
		logger.Log("could not set NoENOBUFS option:", err)
		return
	}

	// Register callback with an error handler
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	err = nf.RegisterWithErrorFunc(ctx, callback, func(e error) int {
		logger.Log("Error:", e)
		return -1
	})
	if err != nil {
		logger.Log("failed to register callback:", err)
		return
	}

	// Set up signal handler to gracefully exit on interrupt
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-sigChan
		logger.Log("Received interrupt signal, shutting down...")
		cancel()
	}()

	<-ctx.Done()
	logger.Log("Exiting...")
}

